### PR TITLE
ruby: specify RUBY_REVISION

### PR DESCRIPTION
**Pull request by thefloweringash**
###### Motivation for this change

Correct the value of `RUBY_REVISION` which is currently defaulting to head `"HEAD"`, to be compatible with [bootsnap's version guards](https://github.com/Shopify/bootsnap/pull/387/files#diff-d1a11f96e76086aab1b9cb2f0f05224d3b2f8fa6e94571d05862c48e54da68a2R11).

```
# Before
❯ nix shell nixpkgs#ruby -c ruby -e 'puts RUBY_REVISION'
HEAD

# After
❯ nix shell nixpkgs#ruby -c ruby -e 'puts RUBY_REVISION'
f69aeb83146be640995753667fdd6c6f157527f5
```

The current build for ruby supports building from git sources and from the upstream tarball. I don't think this is justified, so I simplified to only supporting the upstream tarball, which includes a pregenerated value for `RUBY_REVISION` . The tarball includes a set of gems ("Bundled gems") to be installed by default, which is a change from the current ruby build in nixpkgs.

---

Current Results

| Platform       | Version | Base    | With Rails Express      |
| -------------- | ------- | ------- | ----------------------- |
| aarch64-darwin | 2.7     | ✅      | ✅                      |
| aarch64-darwin | 3.0     | ✅      | ✅                      |
| aarch64-darwin | 3.1     | ✅      | ✅                      |
| x86_64-linux   | 2.7     | ✅      | ✅                      |
| x86_64-linux   | 3.0     | ✅      | ✅                      |
| x86_64-linux   | 3.1     | ✅      | ✅                      |
| x86_64-linux -> aarch64-linux (cross)  | 2.7     | ✅      | ✅                      |
| x86_64-linux -> aarch64-linux (cross)  | 3.0     | ✅      | ✅                      |
| x86_64-linux -> aarch64-linux (cross)  | 3.1     | ❌      | ❌                      |

<details>

<summary>Failure details</summary>

### x86_64-linux -> aarch64-linux, 3.1
```
installing bundled gems:            /nix/store/9gvjx5v5hrmwhy3ja6d87x7fn4d8shg3-ruby-aarch64-unknown-linux-gnu-3.1.0/lib/ruby/gems/3.1.0
                                    minitest 5.15.0
                                    power_assert 2.0.1
                                    rake 13.0.6
                                    test-unit 3.5.3
                                    rexml 3.2.5
                                    rss 0.2.9
                                    net-ftp 0.1.3
                                    net-imap 0.2.2
                                    net-pop 0.1.1
                                    net-smtp 0.3.1
                                    matrix 0.4.2
                                    prime 0.1.2
                                    rbs 2.0.0
Building native extensions. This could take a while...
Traceback (most recent call last):
        18: from ./tool/rbinstall.rb:1124:in `<main>'
        17: from ./tool/rbinstall.rb:1124:in `each'
        16: from ./tool/rbinstall.rb:1127:in `block in <main>'
        15: from ./tool/rbinstall.rb:1043:in `block in <main>'
        14: from ./tool/rbinstall.rb:1043:in `foreach'
        13: from ./tool/rbinstall.rb:1068:in `block (2 levels) in <main>'
        12: from ./tool/rbinstall.rb:899:in `install'
        11: from ./tool/rbinstall.rb:713:in `no_write'
        10: from ./tool/rbinstall.rb:899:in `block in install'
         9: from /build/ruby-3.1.0/lib/rubygems/installer.rb:312:in `install'
         8: from /build/ruby-3.1.0/lib/rubygems/installer.rb:837:in `build_extensions'
         7: from /build/ruby-3.1.0/lib/rubygems/ext/builder.rb:189:in `build_extensions'
         6: from /build/ruby-3.1.0/lib/rubygems/ext/builder.rb:189:in `each'
         5: from /build/ruby-3.1.0/lib/rubygems/ext/builder.rb:192:in `block in build_extensions'
         4: from /build/ruby-3.1.0/lib/rubygems/ext/builder.rb:158:in `build_extension'
         3: from /build/ruby-3.1.0/lib/rubygems/ext/ext_conf_builder.rb:26:in `build'
         2: from /build/ruby-3.1.0/lib/tempfile.rb:317:in `open'
         1: from /build/ruby-3.1.0/lib/rubygems/ext/ext_conf_builder.rb:47:in `block in build'
/build/ruby-3.1.0/lib/rubygems/ext/builder.rb:92:in `run': extconf failed, exit code 1 (Gem::InstallError)
        18: from ./tool/rbinstall.rb:1124:in `<main>'
        17: from ./tool/rbinstall.rb:1124:in `each'
        16: from ./tool/rbinstall.rb:1127:in `block in <main>'
        15: from ./tool/rbinstall.rb:1043:in `block in <main>'
        14: from ./tool/rbinstall.rb:1043:in `foreach'
        13: from ./tool/rbinstall.rb:1068:in `block (2 levels) in <main>'
        12: from ./tool/rbinstall.rb:899:in `install'
        11: from ./tool/rbinstall.rb:713:in `no_write'
        10: from ./tool/rbinstall.rb:899:in `block in install'
         9: from /build/ruby-3.1.0/lib/rubygems/installer.rb:312:in `install'
         8: from /build/ruby-3.1.0/lib/rubygems/installer.rb:837:in `build_extensions'
         7: from /build/ruby-3.1.0/lib/rubygems/ext/builder.rb:189:in `build_extensions'
         6: from /build/ruby-3.1.0/lib/rubygems/ext/builder.rb:189:in `each'
         5: from /build/ruby-3.1.0/lib/rubygems/ext/builder.rb:192:in `block in build_extensions'
         4: from /build/ruby-3.1.0/lib/rubygems/ext/builder.rb:158:in `build_extension'
         3: from /build/ruby-3.1.0/lib/rubygems/ext/ext_conf_builder.rb:26:in `build'
         2: from /build/ruby-3.1.0/lib/tempfile.rb:317:in `open'
         1: from /build/ruby-3.1.0/lib/rubygems/ext/ext_conf_builder.rb:47:in `block in build'
/build/ruby-3.1.0/lib/rubygems/ext/builder.rb:92:in `run': ERROR: Failed to build gem native extension. (Gem::Ext::BuildError)

    current directory: /nix/store/9gvjx5v5hrmwhy3ja6d87x7fn4d8shg3-ruby-aarch64-unknown-linux-gnu-3.1.0/lib/ruby/gems/3.1.0/gems/rbs-2.0.0/ext/rbs_extension
/nix/store/z3mw099xvy75a65a3gc4lcc067lfiva8-ruby-2.7.5/bin/ruby --disable\\=gems -I/build/ruby-3.1.0 -I /build/ruby-3.1.0/lib -r ./siteconf20220314-35968-n2awmt.rb extconf.rb
/build/ruby-3.1.0/rbconfig.rb:13:in `<module:RbConfig>': ruby lib version (3.1.0) doesn't match executable version (2.7.5) (RuntimeError)
        from /build/ruby-3.1.0/rbconfig.rb:11:in `<top (required)>'
        from /nix/store/9gvjx5v5hrmwhy3ja6d87x7fn4d8shg3-ruby-aarch64-unknown-linux-gnu-3.1.0/lib/ruby/gems/3.1.0/gems/rbs-2.0.0/ext/rbs_extension/siteconf20220314-35968-n2awmt.rb:1:in `require'
        from /nix/store/9gvjx5v5hrmwhy3ja6d87x7fn4d8shg3-ruby-aarch64-unknown-linux-gnu-3.1.0/lib/ruby/gems/3.1.0/gems/rbs-2.0.0/ext/rbs_extension/siteconf20220314-35968-n2awmt.rb:1:in `<top (required)>'
        from extconf.rb:in `require'

extconf failed, exit code 1
```

</details>

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/interpreters/ruby/default.nix
Comment (thiagokokada) [inline pkgs/development/interpreters/ruby/default.nix:20]: Could we fallback to the older behavior if `revision` is not set 🤔 ?

Comment (thefloweringash) [reply pkgs/development/interpreters/ruby/default.nix:20]: I don't think that makes sense without more changes. The `generic` function is private to this file, so it will only be called from the section that defines `ruby_2_7`, `ruby_3_0`, and `ruby_3_1`. These callers of the `generic` function should always specify the `revision` of the tag they're fetching.

Ideally the `revision` would be attached to the `src` attribute, so when the `src` was overridden it would use the new value or fall back to `"HEAD"`, but I don't see an easy way to implement this.

File: pkgs/development/interpreters/ruby/default.nix
Comment (thiagokokada) [inline pkgs/development/interpreters/ruby/default.nix:263]: Can you add a comment remembering people to also update this hash when they update the version? Otherwise I can see people forgetting to do this.

Comment (SuperSandro2000) [reply pkgs/development/interpreters/ruby/default.nix:263]: or better we should have an update script. This will definitely be forgotten especially when using nix-update.

**Comment by thefloweringash** [comment]
My initial change was the make the smallest possible change to fix the value of `RUBY_REVISION`. Now that I've had more time to look at this, I think I prefer always building from the tarball which includes a pre-generated value. I don't see any value to building from git, but if I'm missing something important please let me know.

The build failures from the tarball are all related to installing the bundled gems. The current nixpkgs build does not include the bundled gems (they're not included in the git repo). "Bundled" gems are not guaranteed to be present in a ruby install (unlike "Default" gems). If we always build from the tarball and skip the bundled gems, we can simplify the build and get a correct `RUBY_REVISION` from upstream.

File: pkgs/development/interpreters/ruby/default.nix
Comment (thiagokokada) [inline pkgs/development/interpreters/ruby/default.nix:64]: Minor nitpick.

```suggestion
          inherit sha256;
```

File: pkgs/development/interpreters/ruby/patchsets.nix
Comment (thiagokokada) [inline pkgs/development/interpreters/ruby/patchsets.nix:18]: Shouldn't we ship those bundled gems? I mean, we generally should provide an experience as closer as upstream wants, unless we have a better justification.

But maybe this discussion should be made in another issue/PR. I am fine in keeping this here for now.

File: pkgs/development/interpreters/ruby/skip-installing-bundled-gems-2.7.patch
Comment (thiagokokada) [inline pkgs/development/interpreters/ruby/skip-installing-bundled-gems-2.7.patch:17]: Could we set some flag to skip the installation of bundled gems instead of doing this patch? I am thinking in terms of long term maintenance, since each patch we add means that there is a possibility that in the future the patch will fail to apply and someone will have to go there and update the patches.

**Comment by thefloweringash** [comment]
@thiagokokada considering both of your comments: I don't see an easy way to disable the bundled gems, and I don't have a good justification for skipping them other than "the build breaks". I've instead resolved the build issues.

**Review by thiagokokada — COMMENTED** [review]
It mostly looks fine to me, but I would prefer a second opinion before merging this with someone with more experience in Ruby.

Also, given the huge amount of rebuilds, please target this to `staging`.

**Comment by thiagokokada** [comment]
CC @marsam @manveru @vrthra.

**Review by marsam — APPROVED** [review]
LGTM. Thank you!

**Pull request merged**
